src/mpid/ch4/netmod/ofi/ofi_am_impl.h

```diff
@@ -361,11 +396,18 @@ static inline int MPIDI_OFI_do_am_isend(int rank,
          * we should not allocate the entire buffer and do the packing at once. */
         /* TODO: (1) Skip packing for high-density datatypes;
          *       (2) Pipeline allocation for low-density datatypes; */
-        /* FIXME: allocating a GPU registered host buffer adds some additional overhead.
-         * However, once the new buffer pool infrastructure is setup, we would simply be
-         * allocating a buffer from the pool, so whether it's a regular malloc buffer or a GPU
-         * registered buffer should be equivalent with respect to performance. */
-        MPL_gpu_malloc_host((void **) &send_buf, data_sz);
+        if (data_lt_eager_threshold) {
+            MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.am_pack_buf_pool,
+                                               (void **) &send_buf);
+            if (send_buf == NULL) {
+                MPIDI_OFI_deferred_am_isend_enqueue(rank, comm, handler_id, am_hdr_sz, buf, count,
+                                                    datatype, sreq, data_sz);
+                goto fn_exit;
+            }
+        } else {
+            /* only for large message RDMA read */
+            MPL_gpu_malloc_host((void **) &send_buf, data_sz);
```

Comment:
This is needed only for the large message RDMA read at the moment as we do not have a pipelined AM protocol for large messages.

Reply:
We do have it for shared memory (the LMT protocol), just not for network communication.  The RDMA read based protocol makes sense when the data is being sent directly from the user buffer.  But when there's a temporary buffer involved (e.g., for noncontiguous data, or for contiguous data on GPUs when GPU Direct is not available), we should switch to the LMT protocol, as that's more appropriate.

Reply:
I am working on bring the SHM LMT to netmod.